libs/blockscout-auth/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+use cookie::Cookie;
+use std::collections::HashMap;
+use thiserror::Error;
+use tonic::{codegen::http::header::COOKIE, metadata::MetadataMap};
+
+const JWT_TOKEN_NAME: &str = "_explorer_key";
+
+#[derive(Debug)]
+pub struct AuthSuccess {
+    pub user_id: String,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum Error {
+    #[error("invalid jwt token: {0}")]
+    InvalidJwt(String),
+    #[error("invalid cstf token: {0}")]
+    InvalidCsrf(String),
+    #[error("user is unauthorized: {0}")]
+    Unauthorized(String),
+    #[error("blockscout invalid response: {0}")]
+    BlockscoutApi(String),
+}
+
+pub async fn auth_from_metadata(
+    metadata: &MetadataMap,
+    csrf_token: Option<&str>,
+    blockscout_host: &str,
+) -> Result<AuthSuccess, Error> {
+    let jwt = extract_jwt(metadata)?;
+    auth_from_jwt(jwt.as_ref(), csrf_token, blockscout_host).await
+}
+
+pub async fn auth_from_jwt(
+    jwt: &str,
+    csrf_token: Option<&str>,
+    _blockscout_host: &str,
+) -> Result<AuthSuccess, Error> {
+    // TODO: replace with actual blockscout api call
+    let mut jwt = jwt.to_string();
+    if let Some(cstf_token) = csrf_token {
+        jwt.push_str(cstf_token)
+    }
+    Ok(AuthSuccess { user_id: jwt })
```

Comment:
Nit: just some easier option
```suggestion
    let user_id = format!("{jwt}{}", csrf_token.unwrap_or_default());
    Ok(AuthSuccess { user_id } )
```